### DISCUSSION TITLE
OrtNDManager creates PtNDArray

### DISCUSSION THREAD

**Discussion by iprovalo** [deepjavalibrary/djl#2266] [Development]
I have Onnx model I load like so:

```
Criteria<NDList, NDList> criteria = Criteria.builder().setTypes(NDList.class, NDList.class)
    .optEngine("OnnxRuntime").optModelUrls('encoder.onnx').build();

ZooModel<NDList, NDList> encoderModel =  criteria.loadModel();
Predictor<NDList, NDList> encoderPredictor = encoderModel.newPredictor();
OrtNDManager encoderManager = (OrtNDManager) encoderModel.getNDManager();
```

Later in the code I am creating an input tensor:

```
NDManager subManager = encoderManager.newSubManager();
NDArray attention = subManager.create(new long []{1l,2l,3l}).reshape(new Shape(1, 3));
```

When I inspect the `attention` `NDArray` in debug, the `attention` is of type `PtNDArray`, even thought though the `subManager` is `OrtNDManager`.

I am trying to make sure all resources are `Ort` only.  Could you please help me understand how this works?

Thank you,

IP

**Comment by frankfliu** [comment]
@iprovalo 
ONNXRuntime engine can only create NDArray and retrieve the data. It doesn't support NDArray operation. DJL automatically convert to PtNDArray when you calling operators. See Hybrid engine document: https://docs.djl.ai/master/docs/hybrid_engine.html

You can disable alternative engine by setting system property:

```
System.setProperty("ai.djl.onnx.disable_alternative", "true");
```

If alternative engine is disable, any NDArray operations will throw `UnsupportedOperationException`

**Reply by iprovalo** [reply]
Thank you very much @frankfliu for your prompt response!
